encode.go

```diff
@@ -0,0 +1,408 @@
+// Package jsonpatch implements applying and creation of JSON patch as defined in RFC 6902.
+package jsonpatch
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/optiopay/jsonpatch/deep"
+)
+
+var (
+	hyphens = regexp.MustCompile(`[\-_]`)
+)
+
+var (
+	ErrNonPointer     = fmt.Errorf("jsonpatch: interface non-pointer")
+	ErrCouldNotCopy   = fmt.Errorf("jsonpatch: could not make a copy")
+	ErrUnmarshal      = fmt.Errorf("jsonpatch: error while unmarshalling the patch")
+	ErrNodeNil        = fmt.Errorf("jsonpatch: node was empty")
+	ErrIncorrectIndex = fmt.Errorf("jsonpatch: incorrect index")
+	ErrGarbage        = fmt.Errorf("jsonpatch: garbage value")
+	ErrNotImplemented = fmt.Errorf("jsonpatch: not implemented")
+)
+
+type ErrUnsupported struct {
+	Err string
+}
+
+func (e *ErrUnsupported) Error() string {
+	return fmt.Sprintf("jsonpatch: unsupported type for key %s", e.Err)
+}
+
+type patch struct {
+	Op    string          `json:"op"`
+	Path  string          `json:"path"`
+	From  string          `json:"from"`
+	Value json.RawMessage `json:"value"`
+}
```

Comment:
the json tags here are unnecessary since json tags are case insensitive in the go library.  Please be wary of this.